adapters/http4s/src/main/scala/caliban/Http4sAdapter.scala

```diff
@@ -75,6 +107,148 @@ object Http4sAdapter {
       params.get("extensions")
     )
 
+  private def parseGraphQLRequest(body: String): Result[GraphQLRequest] =
+    parse(body).flatMap(_.as[GraphQLRequest]).leftMap(e => DecodingFailure(e.getMessage, Nil))
+
+  private def parsePaths(map: Map[String, Seq[String]]): List[(String, List[Either[String, Int]])] =
+    map.map { case (k, v) => k -> v.map(parsePath).toList }.toList.flatMap(kv => kv._2.map(kv._1 -> _))
+
+  def makeHttpUploadService[R <: Has[_] with Random, E](
+    interpreter: GraphQLInterpreter[R, E],
+    rootUploadPath: Path,
+    blocker: Blocker,
+    skipValidation: Boolean = false,
+    enableIntrospection: Boolean = true,
+    queryExecution: QueryExecution = QueryExecution.Parallel
+  ): HttpRoutes[RIO[R, *]] = {
+    object dsl extends Http4sDsl[RIO[R, *]]
+    import dsl._
+
+    HttpRoutes.of[RIO[R, *]] {
+      case req @ POST -> Root if req.contentType.exists(_.mediaType.isMultipart) =>
```

Comment:
@ghostdogpr hmm, what do you mean by that? I had an idea that because of this extra `R <: Has[_]` constrain, it is easier to put it into a separate method defintion and I see its usage like 

```scala
Http4sAdapter.makeHttpUploadService(...) <+> Http4sAdapter.makeHttpService(...)
```

Reply:
It also makes sense to bake `combineK` into the `makeHttpUploadService` definition but I don't have strong opinions about it 🤷

Reply:
> @ghostdogpr hmm, what do you mean by that? I had an idea that because of this extra `R <: Has[_]` constraint, it is easier to put it into a separate method defintion and I see its usage like 
> 
> 
> 
> ```scala
> 
> Http4sAdapter.makeHttpUploadService(...) <+> Http4sAdapter.makeHttpService(...)
> 
> ```

Yep, that's exactly what I meant. Sounds good 👌